api/address.proto

```diff
@@ -43,6 +43,12 @@ message SocketAddress {
   // should be set for resolution other than DNS. If the address is a concrete
   // IP address, no resolution will occur.
   string resolver_name = 5;
+
+  // When binding to an IPv6 address above, this enables `IPv4 compatibity
+  // <https://tools.ietf.org/html/rfc3493#page-11>`_. Binding to :: will allow
+  // both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into IPv6
+  // space as ::FFFF:<IPv4-address>.
+  bool ipv4_compat = 6;
```

Comment:
Another option would be to allow binding to multiple addresses, and then explicitly specifying both the ipv4 and ipv6 addresses to bind to.  That would also allow binding to multiple ipv4 addresses, for instance.

If we go with this instead, how is this implemented?  Is there a socket option that enables this, or is it implemented by binding two sockets (one for each address family)?

Reply:
I agree multiple addresses are another option. It depends where we want to use this; we have an immediate need in the admin interface, so this would involve adding support for multiple admin addresses. For regular use in listeners, it's already possible to just configure a second listener which is otherwise identical.

For our purposes, the current PR is a small delta that fixes our admin interface problem. If there is broader interest in more flexible admin interface binding from others, I'd be open to doing that instead.

In terms of implementation, this is the `IPV6_V6ONLY` socket option, described in http://man7.org/linux/man-pages/man7/ipv6.7.html. We already set this, forcing it IPv6-only in https://github.com/envoyproxy/envoy/blob/master/source/common/network/address_impl.cc#L223.

Reply:
I think this is fine then.  It is the easiest way to listen on any address for both ipv4 and ipv6.  We can still add multiple-address listeners later if we want/need to.